test/broadcasting.jl

```diff
@@ -39,18 +39,18 @@ end
         @test (df .+ df) ./ 2 !== df
         @test df .+ Matrix(df) == 2 .* df
         @test Matrix(df) .+ df == 2 .* df
-        @test (Matrix(df) .+ df .== 2 .* df) == DataFrame(trues(size(df)), names(df))
+        @test (Matrix(df) .+ df .== 2 .* df) == DataFrame(Tables.table(trues(size(df)), header=names(df)))
```

Comment:
Looks at these examples, it feels unfortunate to lose the symmetry between `Matrix(df)` and `DataFrame(mat)`, especially given the fact that broadcasting between data frames and matrices is allowed.

I think it makes sense to have some asymmetry, since `Matrix(df)` can just drop names, but `DataFrame(mat)` needs to generate names: so requiring `DataFrame(mat, colnames)` isn't illegitimate. But `DataFrame(Tables.table(df, header=colnames))` really breaks that symmetry (and it's really verbose as we see here). Maybe we could keep `DataFrame(mat, colnames)`, since in that case there's no ambiguity (`mat` is treated as a matrix, not as a Table.jl object).

(Sorry for suggesting this now. I can adjust the tests if you like.)

Reply:
I am OK with this (I can fix the tests - it should be quick this way as we have `Tables.table` in all places we need to fix :smile:).

@oxinabox + @pdeffebach : can you comment here please?

The proposal is to have `DataFrame(::AbstractMatrix, ::Union{AbstractVector{Symbol}, AbstractVector{<:AbstractString})` signature (with no default for the second positional argument, so it would be required).

Then there would be no ambiguity with `DataFrame(table)` constructor that can handle matrices as tables.

Reply:
Once again, I stand by my claim that every extra constructor adds confusion.
For this case i think nicer than using Tables.jl is:
```julia
DataFrame(names .=> eachcol(mat))
```
that is a nice clear explict statement about what is going on, and is I think clearer than
`DataFrame(mat, names)` since we are explictly stating that we are going to use the columns one per name.
Also naturally extends to:
`DataFrame(names .=> vec.(eachrow(mat))`
for constructing dataframes based on the rows of the matrix.
Particularly relevent as sometimes we do observations as columns, e.g. Distances.jl, the Models.jl interface, some others.

Reply:
@nalimilan - @oxinabox has convinced me. Writing `DataFrame(names .=> eachcol(mat))` is indeed clean. We have to use `Tables.table` for now because `eachcol` is not supported in Julia 1.0, but once we move the support to new Julia LTS we can clean it up.

Reply:
What I find problematic is that by supporting `Matrix(df)` and `df .+ mat`, we consider that data frames and matrices are objects with compatible shapes. This alone means that we exclude the view of observations as columns (which BTW is the exception in the ecosystem, even in Distances you now have to explicitly give `dims`). Yet `DataFrame(names .=> eachcol(mat))` conceptually requires you to decompose the matrix into columns to create a data frame. The notion that shapes are compatibles is completely lost.

Reply:
The shape is lost, but in `df .+ mat` we explicitly treat `mat` as not being a table (i.e. we assume that it does not have column names). What I mean by this is that `df .+ DataFrame(mat)` currently (i.e. before this PR) would generally fail because column names do not match, so `df .+ mat` does not differ conceptually here from `df .+ vector` or `df .+ scalar` (and we do not expect that `vector` or `scalar` can be converted to a `DataFrame`). In other words:
* broadcasting is about shapes (and that is why it works)
* `DataFrame` constructor is about being a table (which in particular implies having column names)

Reply:
I agree with this. note you can also do `df[1:4, 5:10] .= mat`, which is inconsistent.

Reply:
>     * `DataFrame` constructor is about being a table (which in particular implies having column names)

I don't think this rule holds in general, does it? It's only true for the one-argument constructor (and even in that case, not e.g. for the `::GroupedDataFrame` one). For the two-argument constructor, we don't actually allow tables as the first argument. So it would be OK to accept matrices (just like we accept vectors of vectors).

Reply:
> I don't think this rule holds in general, does it?

We have to have some minimal set of constructors. For instance I hate `DataFrame(a=[1,2], b=[2,3])` constructor as it is inconsistent with the general design we have, but I think we should keep it as it is convenient. How I understand @oxinabox is that he wants to limit the number of special cases as much as possible, and it seems that for matrices `DataFrame(names .=> eachcol(matrix))` is an easy enough pattern.

Reply:
> For instance I hate DataFrame(a=[1,2], b=[2,3]) constructor as it is inconsistent with the general design we have

I was going to suggest deprecating that as well; for `DataFrame(:a=>[1,2], ...)` 😂 
because getting rid of it would mean we are free to add what ever keywords we wanted to the constructor later without clashing with column names.

Reply:
we have this clash already with `copycols` and `makeunique` kwargs which are handled inconsistently:
```
julia> DataFrame(makeunique=2, copycols=false)
1×1 DataFrame
│ Row │ makeunique │
│     │ Int64      │
├─────┼────────────┤
│ 1   │ 2          │
```

Reply:
> I was going to suggest deprecating that as well; for DataFrame(:a=>[1,2], ...)

Oh, is that an option here? I'm 100% on board with that change 😸